nav2_dwb_controller/dwb_controller/src/dwb_controller.cpp

```diff
@@ -124,77 +123,112 @@ DwbController::on_shutdown(const rclcpp_lifecycle::State &)
-
+    setPlannerPath(action_server_->get_current_goal()->path);
     ProgressChecker progress_checker(rclcpp_node_);
 
-    rclcpp::Rate loop_rate(10);
+    rclcpp::Rate loop_rate(100ms);
     while (rclcpp::ok()) {
-      nav_2d_msgs::msg::Pose2DStamped pose2d;
-      if (!getRobotPose(pose2d)) {
-        RCLCPP_INFO(get_logger(), "No pose. Stopping the robot");
+      if (shouldCancel()) {
+        return;
+      }
+
+      checkPreemption();
+
+      if (!computeAndUpdate(progress_checker)) {
+        RCLCPP_WARN(get_logger(), "Error during control computation. Stopping the robot");
         publishZeroVelocity();
-      } else {
-        if (isGoalReached(pose2d)) {
-          RCLCPP_INFO(get_logger(), "Reached the goal");
-          break;
-        }
-        progress_checker.check(pose2d);
-        auto velocity = odom_sub_->getTwist();
-        auto cmd_vel_2d = planner_->computeVelocityCommands(pose2d, velocity);
-
-        RCLCPP_DEBUG(get_logger(), "Publishing velocity at time %.2f", now().seconds());
-        publishVelocity(cmd_vel_2d);
-
-        if (current_goal_handle->is_canceling()) {
-          RCLCPP_INFO(this->get_logger(), "Canceling execution of the local planner");
-          current_goal_handle->canceled(result);
-          publishZeroVelocity();
-          return;
-        }
-
-        // Check if there is an update to the path to follow
-        if (action_server_->preempt_requested()) {
-          RCLCPP_INFO(get_logger(), "Received a new goal, pre-empting the old one");
-          current_goal_handle = action_server_->get_updated_goal_handle();
-          goal = current_goal_handle->get_goal();
-
-          // If so, pass the new path to the local planner
-          auto path = nav_2d_utils::pathToPath2D(goal->path);
-          planner_->setPlan(path);
-        }
+      }
+
+      if (isGoalReached()) {
+        RCLCPP_INFO(get_logger(), "Reached the goal!");
+        break;
       }
 
       loop_rate.sleep();
     }
   } catch (nav_core2::PlannerException & e) {
     RCLCPP_ERROR(this->get_logger(), e.what());
     publishZeroVelocity();
-    current_goal_handle->abort(result);
+    action_server_->abort_all();
     return;
   }
 
   RCLCPP_DEBUG(get_logger(), "DWB succeeded, setting result");
-  current_goal_handle->succeed(result);
+
+  // TODO(orduno) #861 Handle a pending preemption.
+  action_server_->succeeded_current();
+
   publishZeroVelocity();
 }
 
+void DwbController::setPlannerPath(const nav2_msgs::msg::Path & path)
+{
+  auto path2d = nav_2d_utils::pathToPath2D(path);
+
+  RCLCPP_DEBUG(get_logger(), "Providing path to the local planner");
+  planner_->setPlan(path2d);
+
+  auto end_pose = *(path.poses.end() - 1);
+
+  RCLCPP_DEBUG(get_logger(), "Path end point is (%.2f, %.2f)",
+    end_pose.position.x, end_pose.position.y);
+}
+
+bool DwbController::computeAndUpdate(ProgressChecker & progress_checker)
+{
+  nav_2d_msgs::msg::Pose2DStamped pose2d;
+
+  if (!getRobotPose(pose2d)) {
+    return false;
```

Comment:
In the old code, this would trigger a call to `publish_zero_velocity`. Rather than add that back in here, I'm thinking it would make sense to just throw an exception, have DWB fail and publish_zero_velocity as part of the exception handler.

Reply:
Agree, will change that.